app/controllers/locationOfGoods/IdentificationController.scala

```diff
@@ -55,46 +55,70 @@ class IdentificationController @Inject() (
+                  case None        => form(identifiers)
+                  case Some(value) => form(identifiers).fill(value)
+                }
 
-            Future.successful(Ok(view(preparedForm, departureId, identifiers, mode)))
+                Future.successful(Ok(view(preparedForm, departureId, identifiers, mode)))
+            }
+          case None => Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
         }
     }
 
   def onSubmit(departureId: String, mode: Mode): Action[AnyContent] = actions
     .requireData(departureId)
-    .andThen(getMandatoryPage(LocationTypePage, InferredLocationTypePage))
     .async {
       implicit request =>
-        locationOfGoodsIdentificationTypeService.getLocationOfGoodsIdentificationTypes(request.arg).flatMap {
-          locationOfGoodsIdentificationTypes =>
-            form(locationOfGoodsIdentificationTypes)
-              .bindFromRequest()
-              .fold(
-                formWithErrors => Future.successful(BadRequest(view(formWithErrors, departureId, locationOfGoodsIdentificationTypes, mode))),
-                value => redirect(mode, IdentificationPage, value, departureId)
-              )
+        getLocationType match {
+          case Some(location) =>
+            locationOfGoodsIdentificationTypeService.getLocationOfGoodsIdentificationTypes(location).flatMap {
+              locationOfGoodsIdentificationTypes =>
+                form(locationOfGoodsIdentificationTypes)
+                  .bindFromRequest()
+                  .fold(
+                    formWithErrors => Future.successful(BadRequest(view(formWithErrors, departureId, locationOfGoodsIdentificationTypes, mode))),
+                    value => redirect(mode, IdentificationPage, value, departureId)
+                  )
+            }
+          case None => Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
         }
+
     }
 
   private def redirect(
     mode: Mode,
     page: QuestionPage[LocationOfGoodsIdentification],
     value: LocationOfGoodsIdentification,
     departureId: String
-  )(implicit request: Request): Future[Result] =
+  )(implicit request: MandatoryDataRequest[_]): Future[Result] =
     for {
       updatedAnswers <- Future.fromTry(request.userAnswers.set(page, value))
       _              <- sessionRepository.set(updatedAnswers)
     } yield Redirect(navigator.nextPage(page, updatedAnswers, departureId, mode))
+
+  private def getLocationType(implicit request: DataRequest[AnyContent]): Option[String] = {
+    val getLocationType: Option[String] = request.userAnswers
+      .get(LocationTypePage)
+      .map(_.`type`)
+      .orElse(
+        request.userAnswers
+          .get(InferredLocationTypePage)
+          .map(_.`type`)
+      )
+      .orElse(
+        request.userAnswers.departureData.Consignment.LocationOfGoods.map(
+          x => x.typeOfLocation
+        )
+      )
+    getLocationType
```

Comment:
Do we need to assign this to a val if we're just returning it?

Reply:
intelij refactor method's idea, removed.